addons/survey/models/survey_user.py

```diff
@@ -267,240 +267,96 @@ def write(self, vals):
 
     @api.model
-    def save_line_numerical_box(self, user_input_id, question, post, answer_tag):
-        vals = {
-            'user_input_id': user_input_id,
-            'question_id': question.id,
-            'survey_id': question.survey_id.id,
-            'skipped': False
-        }
-        if answer_tag in post and post[answer_tag].strip():
-            vals.update({'answer_type': 'number', 'value_number': float(post[answer_tag])})
+    def _save_line_simple_answer(self, vals, old_answers, question, answer):
+        answer_type = question.question_type
+        if question.question_type == 'textbox':
+            answer_type = 'text'
+        elif question.question_type == 'numerical_box':
+            answer_type = 'number'
+
+        vals.update(self._get_save_line_values(answer, answer_type))
+        if old_answers:
+            old_answers.write(vals)
         else:
-            vals.update({'answer_type': None, 'skipped': True})
-        old_uil = self.search([
-            ('user_input_id', '=', user_input_id),
-            ('survey_id', '=', question.survey_id.id),
-            ('question_id', '=', question.id)
-        ])
-        if old_uil:
-            old_uil.write(vals)
-        else:
-            old_uil.create(vals)
+            self.create(vals)
         return True
 
     @api.model
-    def save_line_date(self, user_input_id, question, post, answer_tag):
-        vals = {
-            'user_input_id': user_input_id,
-            'question_id': question.id,
-            'survey_id': question.survey_id.id,
-            'skipped': False
-        }
-        if answer_tag in post and post[answer_tag].strip():
-            vals.update({'answer_type': 'date', 'value_date': post[answer_tag]})
-        else:
-            vals.update({'answer_type': None, 'skipped': True})
-        old_uil = self.search([
-            ('user_input_id', '=', user_input_id),
-            ('survey_id', '=', question.survey_id.id),
-            ('question_id', '=', question.id)
-        ])
-        if old_uil:
-            old_uil.write(vals)
-        else:
-            old_uil.create(vals)
-        return True
+    def _save_line_choice(self, vals, old_answers, question, answers, comment):
+        if not (isinstance(answers, list)):
+            answers = [answers]
+
+        vals_list = []
+        if question.question_type == 'simple_choice':
+            if not (question.comment_count_as_answer and question.comments_allowed and comment):
+                for answer in answers:
+                    vals.update(self._get_save_line_values(answer, 'suggestion'))
+                    vals_list.append(vals)
+        elif question.question_type == 'multiple_choice':
+            for answer in answers:
+                vals.update(self._get_save_line_values(answer, 'suggestion'))
+                vals_list.append(vals.copy())
+
+        if comment:
```

Comment:
Don't you have to check that `question.comment_count_as_answer and question.comments_allowed` too ? Not sure.